src/snovault/elasticsearch/simple_queue.py

```diff
@@ -4,154 +4,206 @@
         '''
-        if self.has_uuids():
+        if self.is_indexing():
             return None
         for uuid in uuids:
             self._load_uuid(uuid)
         return len(self._uuids)
 
-    def reset(self, force=False):
-        '''
-        Reset the queue so a new indexing process can begin
-        - There cannot be any uuids remaining
-        - The errors from the last indexing session must be removed
-        - There cannot be any pending client connections
-        '''
-        msg = 'Okay'
-        can_reset = True
-        if self.has_uuids():
-            can_reset = False
-            msg = 'Queue Server cannot reset. Uuids=%d' % (
-                len(self._uuids),
-            )
-        if self.errors:
-            can_reset = False
-            msg = 'Queue Server cannot reset. Errors=%d' % (
-                len(self.errors),
-            )
-        for client_id, client_info in self._client_conns.items():
-            if client_info['uuid_cnt']:
-                can_reset = False
-                msg = 'Queue Server cannot reset. Client Pending(%s)' % (
-                    client_id,
-                )
-        if can_reset or force:
-            self._uuids = []
-            self._client_conns = {}
-            self.errors = []
-        return msg
+    # Run
+    def is_indexing(self):
+        '''Is an indexing process currently running'''
+        if self.has_uuids() or self._has_errors():
+            return True
+        for worker_conn in self._worker_conns.values():
+            if worker_conn['uuid_cnt']:
+                return True
+        return False
 
-    def update_finished(self, client_id, results):
-        '''The way a client finishes a batch of uuids'''
-        msg = 'Client id(%s) DNE.' % client_id
-        if client_id in self._client_conns:
-            client_conn = self._client_conns[client_id]
-            client_conn['results'] = results
-            batch_errors = []
-            for error in results['errors']:
-                error['client_id'] = client_id
-                batch_errors.append(error)
-            self.errors.extend(batch_errors)
-            client_conn['uuid_cnt'] -= results['successes']
-            client_conn['uuid_cnt'] -= len(batch_errors)
-            if results['successes'] < 0 or client_conn['uuid_cnt']:
-                msg = 'Queue Server cannot close client(%s)' % client_id
-            else:
-                msg = 'Okay'
+    def update_finished(self, given_worker_id, given_results):
+        '''The way a worker finishes a batch of uuids'''
+        msg = 'Worker id(%s) DNE.' % given_worker_id
+        worker_conns = self.get_worker_conns()
+        for worker_id, worker_conn in worker_conns.items():
+            if worker_id == given_worker_id:
+                errors_added = self.add_errors(worker_id, given_results['errors'])
+                uuid_cnt = worker_conn['uuid_cnt'] - given_results['successes'] - errors_added
```

Comment:
Will this ever be less than zero if there is multiple errors?

Reply:
Checked.  Update object will only return one error per uuid.